onenetd.1
.TH onenetd 1 "19th June 2014" "Version 12" "Network Tools"
.SH NAME
onenetd \- listen on a TCP port and launch a program to handle requests
.SH SYNOPSIS
onenetd [options] ADDRESS PORT COMMAND ...
.SH DESCRIPTION
onenetd binds a TCP socket to port PORT on IP address ADDRESS, listens for and
accepts connections, and spawns COMMAND with stdin and stdout (and optionally
stderr) connected to the accepted socket.
.PP
onenetd may be thought of as a minimal version of inetd, suitable only for
listening for "stream tcp nowait" services on a single port. The idea is that
multiple copies of onenetd will be run, one for each service. The advantages of
onenetd over inetd include:
.IP \(bu
Connections may be refused with a fixed response after the connection limit is
reached.
.IP \(bu
TCP_NODELAY may be set on a per-port basis.
.IP \(bu
The address to bind to may be chosen arbitrarily, so it's easy to make a
service only listen on localhost, or run several different services on the same
port for different interface addresses.
.IP \(bu
Incoming connections may be logged to stderr.
.PP
Note that onenetd only handles TCP connections; it is not possible to run UDP
services (such as nmbd) from onenetd.
.SH OPTIONS
.TP
.B ADDRESS
The hostname or IP address to bind() to. Specify 0 (or :: for IPv6) to bind to
all addresses.
.TP
.B PORT
The service name or port number to bind() to. In order to bind to a privileged
port, onenetd must be initially run as root (although you can use the -u etc.
options to switch to another user after the bind has been done). Specify 0 to
bind to any available port.
.TP
.B COMMAND ...
The command to execute, as well as its parameters (all parameters given to
onenetd after COMMAND are passed to COMMAND as parameters). argv[0] for COMMAND
will be set to COMMAND; if you wish to specify a different argv[0], use the
argv0 tool from the freedt or daemontools packages.
.TP
.B \-c N
Limit the number of connections (and thus child processes) to N. The default is
40.
.TP
.B \-6
Bind to an IPv6 address. IPv6 sockets will also accept IPv4 connections, if
bound to an appropriate address (e.g. ::).
.TP
.B \-g GID
After binding the TCP socket, setgid(GID). GID must be numeric.
.TP
.B \-u UID
After binding the TCP socket, setuid(UID). UID must be numeric.
.TP
.B \-U
After binding the TCP socket, setgid($GID) and setuid($UID). $GID and $UID must
be numeric. This is intended for use with envuidgid (from daemontools or
freedt).
.TP
.B \-1
After binding the TCP socket, print the local port number to stdout.
You can use this when you've specified the local port as 0 to find out what
port has been chosen.
.TP
.B \-b N
Set the listen() backlog to N. Note that on many operating systems, the listen
backlog has an arbitrary limit, or may be entirely ignored. The default is 10.
.TP
.B \-D
Set the TCP_NODELAY option on the accepted sockets. This causes the operating
system's TCP stack to avoid coalescing smaller packets into larger ones,
decreasing latency but reducing throughput.
.TP
.B \-Q
Do not print any messages to stderr, except in the case of a fatal error. This
is the default.
.TP
.B \-v
Print messages to stderr indicating clients connecting and disconnecting.
These can be collected for logging purposes using logger, multilog from
daemontools, or dumblog from freedt.
.TP
.B \-e
Redirect the child's stderr to the socket. By default, only stdin and stdout
are redirected.
.TP
.B \-h
Print a brief usage message, then exit. The same message will be shown if
unknown or invalid arguments are passed.
.TP
.B \-r MESSAGE
Normally once the number of active connections has passed the limit set by -c,
further connections will not be accepted until the number of active connections
has dropped again -- that is, further connections will not be refused, but will
not connect until free connections are available. If -r is specified,
connections will be accepted and have the constant string MESSAGE printed to
them, then will be disconnected; this is done inside the onenetd process
without forking, so it does not use up process slots. MESSAGE may contain \\r,
\\n or \\t to specify a carriage return, newline or tab character.
.SH EXIT CODE
onenetd returns 0 on success, or 20 on failure (for instance, invalid
arguments). Since onenetd is intended to run forever, the exit code is
generally not important, and stderr should be examined for diagnostic messages
if it exits unexpectedly.
.SH ENVIRONMENT VARIABLES
Following D. J. Bernstein's UCSPI standard, onenetd sets several variables in
the child process's environment. These can be used for logging or access
control.
.TP
.B PROTO
The string "TCP" for IPv4 connections, or "TCP6" for IPv6 connections.
.TP
.B TCPLOCALIP
The local address of the connected socket, as formatted by inet_ntop (dotted
quad for IPv4, hex for IPv6).
.TP
.B TCPLOCALPORT
The local port of the connected socket, as a decimal number.
.TP
.B TCPREMOTEIP
The remote address of the connected socket, as formatted by inet_ntop.
.TP
.B TCPREMOTEPORT
The remote port of the connected socket, as a decimal number.
.SH EXAMPLES
.TP
onenetd 0 echo cat
Simulate the standard inetd echo service.
.TP
onenetd -v 192.168.1.2 daytime date
Simulate the standard inetd daytime service on the interface 192.168.1.2,
logging connections to stderr.
.TP
onenetd -1 ::1 0 myprog --my-arguments
Bind to a free port on the IPv6 localhost address, print the port number
chosen, and run "myprog --my-arguments" for connections.
.TP
onenetd -v -c 5 -r '421 Server full\\r\\n' 0 21 /usr/sbin/ftpd -a /home/ftp
Run an FTP service with a limit of five concurrent connections, refusing any
further connections with an appropriate FTP error message. ftpd will be invoked
as "/usr/sbin/ftpd -a /home/ftp".
.TP
envuidgid myuser onenetd -v -U 0 25098 myprog --my-arguments
Set user and group to that of myuser, and run "myprog --my-arguments" for
connections to port 25098 on any IPv4 interface.
.SH AUTHOR
onenetd was written by Adam Sampson <ats@offog.org>.
.SH SEE ALSO
.BR inetd (8),
.BR argv0 (8),
.BR envuidgid (8),
.BR softlimit (8),
.BR dumblog (8),
.BR logger (1),
.BR sock (1).


NEWS
- onenetd 1.12

Tidy up the documentation: remove duplicate copyright notices, merge
the ChangeLog and NEWS files together, ensure all options and
environment variables are documented in the help and man page.

Change the license (again) from the Expat license to the ISC license,
which is equivalent and simpler.

Refactor the code in the interests of making the main function a bit
less huge and deeply nested.

Make onenetd work on ARM: correct a number of type errors in the code,
where it was using an inappropriate type to hold a return value from a
library function.

Fix an uninitialised variable use (spotted by clang).

Check that strdup hasn't failed when setting up the environment (spotted
by clang).

Add the -1 option.

Automate the test suite. There used to be some scripts in the repository
for testing it by hand; now "make check" will do the right thing.

Add IPv6 support. This behaves like Felix von Leitner's IPv6 patch for
tcpserver: specify -6, and onenetd will bind to an IPv6 socket. If
you've specified an address that can accept IPv4 connections (e.g. ::),
it'll work for both.

- onenetd 1.11

Change the license from the GPL to the Expat license.

Use getsockname() to get the local address to put in TCPLOCALIP, rather
than just using the listen address; this means that TCPLOCALIP will be
set to the address of the interface the client connected to.

Document the environment variables. Reported by Tim Bishop.

Check that accept() returned a valid address.

Catch a (harmless) out-of-memory crash while parsing options.  Reported
by Ulf Härnhammar, who also did a security review of the code -- thanks!

Update the list of connected clients correctly; previously
clients weren't being added to the list at all, so memory would be
leaked if the refusal message wasn't accepted immediately, and the child
FD would be closed immediately anyway. Avoid overfilling fd_sets by
dropping client connections if they wouldn't fit.

- onenetd 1.10

Set fd options correctly (required for Solaris).

Block SIGCHLD except while select() is running.

Use a self-pipe again, since the other way is not correct (it leaves
zombies if the handler gets called early enough).

Use FD_CLOEXEC to avoid leaking fds.

Various cleanups.

- onenetd 1.9

Made all the errors that can occur in the main loop non-fatal.

- onenetd 1.8

Added -e to redirect stderr of child to socket.

Removed obsolete -o and -d from getopt args.

Wrote manual page.

- onenetd 1.7

Fixed bug where onenetd wouldn't accept any connections after it had
been full -- it was calculating "full" in the wrong place.

Made the disconnect message show the connection count after
disconnection, which makes it easier to see how many clients are
actually connected from the log.

- onenetd 1.6

Avoid selecting on the listening socket when full.

Split out the "send response" code, so that we can call it when we first
notice a socket needs a response, which saves a call to select().

Made the status messages include the number of children running.

- onenetd 1.5

Removed redundant test for host being "0", since inet_addr handles this
case already.

Changed uid and gid to longs, since uid_t and gid_t aren't signed under
dietlibc.

Made help a bit more helpful.

Removed malloc debugging code.

- onenetd 1.4

Fix linking under Solaris.

- onenetd 1.3

Due to missing features on Solaris: removed -K (kill IP options),
replaced setenv with putenv, replaced inet_ntoa with inet_addr.

Added escape characters in response string.

- onenetd 1.2

Added escape characters in response.

- onenetd 1.1

Initial release.


onenetd.c
/*
   onenetd: a single-process inetd equivalent
   Copyright 2001, 2002, 2003, 2005, 2014 Adam Sampson <ats@offog.org>

   Permission to use, copy, modify, and/or distribute this software for any
   purpose with or without fee is hereby granted, provided that the above
   copyright notice and this permission notice appear in all copies.

   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include <sys/types.h>
#include <signal.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <stdio.h>
#include "config.h"

int max_conns = 40;
int conn_count = 0;
int bind_family = AF_INET;
int use_gid = 0;
gid_t gid = 0;
int use_uid = 0;
uid_t uid = 0;
int show_port = 0;
int backlog = 10;
int no_delay = 0;
int verbose = 0;
int stderr_to_socket = 0;
char *response = NULL;
char **command;

/* This pipe is used to safely detect SIGCHLD: the SIGCHLD handler writes a
   character to it, and the main loop can then reap children later.
   (See http://cr.yp.to/docs/selfpipe.html for details.) */
int selfpipe[2];

typedef struct client {
	int fd;
	char *message;
	size_t left;
	struct client *next;
} client;
client *clients = NULL;

/* Structure big enough to contain either an IPv4 or IPv6 socket address. */
typedef union {
	struct sockaddr_in v4;
	struct sockaddr_in6 v6;
} either_addr_t;

/* Get the UCSPI PROTO value for an address. */
const char *get_proto(const either_addr_t *addr) {
	if (addr->v4.sin_family == AF_INET6
		&& !IN6_IS_ADDR_V4MAPPED(&addr->v6.sin6_addr))
		return "TCP6";
	else
		return "TCP";
}

/* Get the UCSPI *IP value for an address.
   Returns a pointer to a static buffer. */
const char *get_addr(const either_addr_t *addr) {
	static char buf[INET6_ADDRSTRLEN];
	const void *src;
	int family = addr->v4.sin_family;

	if (family == AF_INET) {
		src = &addr->v4.sin_addr;
	} else if (IN6_IS_ADDR_V4MAPPED(&addr->v6.sin6_addr)) {
		/* An IPv4-mapped IPv6 address; display as IPv4. */
		family = AF_INET;
		src = ((char *)(&addr->v6.sin6_addr)) + 12;
	} else {
		src = &addr->v6.sin6_addr;
	}

	const char *s = inet_ntop(family, src, buf, sizeof buf);
	if (s == NULL)
		return "-";
	return s;
}

/* Get the UCSPI *PORT value for an address. */
int get_port(const either_addr_t *addr) {
	if (addr->v4.sin_family == AF_INET)
		return ntohs(addr->v4.sin_port);
	else
		return ntohs(addr->v6.sin6_port);
}

/* Print a warning. */
void warn(const char *msg) {
	fprintf(stderr, "%s\n", msg);
}

/* Die with an error message. */
void die(const char *msg) {
	warn(msg);
	exit(20);
}

/* Handle SIGCHLD. */
void handle_sigchld(int dummy) {
	int old_errno = errno;
	write(selfpipe[1], "c", 1);
	errno = old_errno;
}

/* Change the flags on an fd. */
int change_flags(int fd, int add, int remove) {
	int flags = fcntl(fd, F_GETFL);
	int newflags;
	if (flags == -1)
		return -1;

	newflags = (flags | add) & ~remove;

	if (newflags != flags) {
		if (fcntl(fd, F_SETFL, flags) < 0)
			return -1;
	}
	
	return 0;
}

/* Set the FD_CLOEXEC flag on an fd. */
void set_fd_cloexec(int fd) {
	if (fcntl(fd, F_SETFD, FD_CLOEXEC) < 0)
		die("unable to set FD_CLOEXEC");
}

/* Add an fd to an FD_SET, updating a maximum. */
void fd_set_add(int fd, fd_set *fds, int *max) {
	FD_SET(fd, fds);
	if (fd > *max) *max = fd;
}

/* Equivalent to putenv(strdup(s)), with error checking. */
int putenv_dup(const char *s) {
	char *copy = strdup(s);
	if (copy == NULL)
		die("strdup failed");

	return putenv(copy);
}

/* Print the usage message. */
void usage(int code) {
	fprintf(stderr, "onenetd version " VERSION "\n"
		"\n"
		"Usage: onenetd [options] address port command ...\n"
		"  address  Address to bind to (0 for all local addresses)\n"
		"  port     TCP port to bind to (0 for any available port)\n"
		"  command  Command to execute\n"
		"Options:\n"
		"  -c N     limit to at most N children running (default 40).\n"
		"           Further connections will be deferred unless -r\n"
		"           is specified.\n"
		"  -6       bind to an IPv6 address (default IPv4)\n"
		"  -g gid   setgid(gid) after binding\n"
		"  -u uid   setuid(uid) after binding\n"
		"  -U       setuid($UID) and setgid($GID) after binding\n"
		"  -1       print local port number to stdout after binding\n"
		"  -b N     set listen() backlog to N\n"
		"  -D       set TCP_NODELAY option on sockets\n"
		"  -e       redirect stderr of children to socket\n"
		"  -v       be verbose\n"
		"  -Q       don't be verbose (default)\n"
		"  -r resp  once -c limit is reached, refuse clients\n"
		"           with 'resp' rather than deferring them.\n"
		"           resp may contain \\r, \\n, \\t.\n"
		"  -h       show this usage message\n"
		"\n"
		"Report bugs to <ats@offog.org>.\n");
	exit(code);
}

/* Create and bind the listening socket. */
int make_listen_socket(const char *address, const char *port) {
	struct addrinfo hints = {};
	struct addrinfo *ai;
	int rc, n, fd;

	hints.ai_family = bind_family;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_flags = AI_PASSIVE;
	hints.ai_protocol = 0;

	rc = getaddrinfo(address, port, &hints, &ai);
	if (rc != 0) {
		die(gai_strerror(rc));
	}

	fd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
	if (fd < 0)
		die("unable to create socket");
	if (change_flags(fd, O_NONBLOCK, 0) < 0)
		die("unable to set O_NONBLOCK");
	set_fd_cloexec(fd);
	n = 1;
	if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &n, sizeof n) < 0)
		die("unable to set SO_REUSEADDR");
	if (bind(fd, ai->ai_addr, ai->ai_addrlen) < 0)
		die("unable to bind to listen address");
	if (listen(fd, backlog) < 0)
		die("unable to listen");

	if (show_port) {
		either_addr_t addr;
		socklen_t size = sizeof addr;

		if (getsockname(fd, (struct sockaddr *)&addr, &size) < 0)
			die("unable to get bound address");

		printf("%d\n", get_port(&addr));
		fflush(stdout);
	}

	freeaddrinfo(ai);

	return fd;
}

/* Try to send a chunk of the response to a client. Remove the client
   from the list if we've sent all of it. */
void try_to_send(client *prev_cl, client *cl) {
	int remove = 0;
	ssize_t count = write(cl->fd, cl->message, cl->left);

	if (count >= 0) {
		cl->message += count;
		cl->left -= count;

		if (cl->left == 0)
			remove = 1;
	} else if (errno == EAGAIN) {
		/* ignorable error */
	} else {
		/* another error while writing */
		remove = 1;
	}

	if (remove) {
		close(cl->fd);
		if (prev_cl) {
			prev_cl->next = cl->next;
		} else {
			clients = cl->next;
		}
		free(cl);
	}
}

/* Accept a new connection, and either spawn a new child process or add it to
   the list of clients to reject. */
void accept_connection(int listen_fd, int full) {
	pid_t pid;
	either_addr_t local_addr, child_addr;
	socklen_t len = sizeof child_addr;
	int child_fd;
	int n;

	child_fd = accept(listen_fd, (struct sockaddr *)&child_addr, &len);
	if (len > sizeof child_addr) {
		warn("unable to get remote address");
		goto no_conn;
	}
	if (child_fd < 0 && errno == EAGAIN)
		goto no_conn;
	if (child_fd < 0) {
		warn("accept failed");
		goto no_conn;
	}
	set_fd_cloexec(child_fd);

	len = sizeof local_addr;
	if (getsockname(child_fd, (struct sockaddr *)&local_addr, &len) < 0
		|| len > sizeof local_addr) {
		warn("unable to get local address");
		goto no_conn;
	}

	if (full) {
		client *cl;

		/* Avoid overfilling the fd_set. */
		if (child_fd >= FD_SETSIZE && verbose) {
			fprintf(stderr, "- dropped from %s port %d\n",
				get_addr(&child_addr),
				get_port(&child_addr));
		}
		if (child_fd >= FD_SETSIZE)
			goto no_conn;

		if (change_flags(child_fd, O_NONBLOCK, 0) < 0) {
			warn("unable to set O_NONBLOCK");
			goto no_conn;
		}

		cl = malloc(sizeof *cl);
		if (!cl) {
			warn("out of memory");
			goto no_conn;
		}

		cl->fd = child_fd;
		child_fd = -1;
		cl->message = response;
		cl->left = strlen(cl->message);
		cl->next = clients;
		clients = cl;

		if (verbose)
			fprintf(stderr, "- refused from %s port %d\n",
				get_addr(&child_addr),
				get_port(&child_addr));

		/* Try to send the response now; if we send
		   all of it it'll get removed from the list
		   again. */
		try_to_send(NULL, cl);

		goto no_conn;
	}

	n = 1;
	if (no_delay && setsockopt(child_fd, IPPROTO_TCP,
		TCP_NODELAY, &n, sizeof n) < 0) {
		warn("unable to set TCP_NODELAY");
		goto no_conn;
	}

	pid = fork();
	if (pid < 0) {
		warn("fork failed");
		goto no_conn;
	}
	if (pid == 0) {
		char buf[80];

		dup2(child_fd, 0);
		dup2(child_fd, 1);
		if (stderr_to_socket)
			dup2(child_fd, 2);

		snprintf(buf, sizeof buf, "PROTO=%s",
			get_proto(&local_addr));
		putenv_dup(buf);
		snprintf(buf, sizeof buf, "TCPLOCALIP=%s",
			get_addr(&local_addr));
		putenv_dup(buf);
		snprintf(buf, sizeof buf, "TCPLOCALPORT=%d",
			get_port(&local_addr));
		putenv_dup(buf);
		snprintf(buf, sizeof buf, "TCPREMOTEIP=%s",
			get_addr(&child_addr));
		putenv_dup(buf);
		snprintf(buf, sizeof buf, "TCPREMOTEPORT=%d",
			get_port(&child_addr));
		putenv_dup(buf);

		execvp(command[0], command);
		_exit(20);
	}

	conn_count++;
	if (verbose)
		fprintf(stderr, "%ld connected from %s port %d (%d/%d)\n",
			(long) pid,
			get_addr(&child_addr),
			get_port(&child_addr),
			conn_count, max_conns);

no_conn:
	if (child_fd >= 0)
		close(child_fd);
}

/* Check for child processes that have exited. */
void reap_children(void) {
	while (1) {
		pid_t pid = waitpid(-1, NULL, WNOHANG);
		if (pid <= 0)
			break;

		conn_count--;
		if (verbose)
			fprintf(stderr, "%ld closed (%d/%d)\n",
				(long) pid, conn_count, max_conns);
	}
}

int main(int argc, char **argv) {
	struct sigaction sa;
	sigset_t sig_chld;
	int listen_fd;
	char *s, *r;
	int n;

	while (1) {
		int c = getopt(argc, argv, "+c:6g:u:U1b:DQvehr:");
		if (c == -1)
			break;
		switch (c) {
		case 'c':
			max_conns = atoi(optarg);
			break;
		case '6':
			bind_family = AF_INET6;
			break;
		case 'g':
			use_gid = 1;
			gid = atoi(optarg);
			break;
		case 'u':
			use_uid = 1;
			uid = atoi(optarg);
			break;
		case 'U':
			s = getenv("GID");
			if (!s)
				die("-U specified but no $GID");
			use_gid = 1;
			gid = atoi(s);
			s = getenv("UID");
			if (!s)
				die("-U specified but no $UID");
			use_uid = 1;
			uid = atoi(s);
			break;
		case '1':
			show_port = 1;
			break;
		case 'b':
			backlog = atoi(optarg);
			break;
		case 'D':
			no_delay = 1;
			break;
		case 'Q':
			verbose = 0;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'e':
			stderr_to_socket = 1;
			break;
		case 'h':
			usage(0);
			break;
		case 'r':
			r = response = malloc(strlen(optarg) + 1);
			if (!r)
				die("out of memory");
			for (s = optarg; *s != '\0'; s++) {
				if (*s == '\\') {
					s++;
					if (*s == 'r')
						*r++ = '\r';
					else if (*s == 'n')
						*r++ = '\n';
					else if (*s == 't')
						*r++ = '\t';
					else
						usage(20);
				} else {
					*r++ = *s;
				}
			}
			*r = '\0';
			break;
		default:
			usage(20);
		}
	}

	if ((argc - optind) < 3)
		usage(20);

	listen_fd = make_listen_socket(argv[optind], argv[optind + 1]);
	command = &argv[optind + 2];

	/* Drop privileges. */
	if (use_gid)
		if (setgid(gid) < 0)
			die("unable to setgid");
	if (use_uid)
		if (setuid(uid) < 0)
			die("unable to setuid");

	/* Create the self-pipe. */
	if (pipe(selfpipe) < 0)
		die("unable to create self-pipe");
	if (change_flags(selfpipe[1], O_NONBLOCK, 0) < 0)
		die("unable to set O_NONBLOCK");
	set_fd_cloexec(selfpipe[0]);
	set_fd_cloexec(selfpipe[1]);

        /* Mask SIGCHLD, except when we're blocked in select(). This is because
           many of the system calls we use are interruptable, and we'd
           otherwise have to handle EINTR everywhere.  (It would be simpler to
           just use SA_RESTART for SIGCHLD -- but POSIX says that it's
           implementation-defined whether select() is interrupted in that case
           or not.) */
	sigemptyset(&sig_chld);
	sigaddset(&sig_chld, SIGCHLD);
	if (sigprocmask(SIG_BLOCK, &sig_chld, NULL) < 0)
		die("unable to block SIGCHLD");

	sa.sa_handler = handle_sigchld;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = SA_NOCLDSTOP;
	sigaction(SIGCHLD, &sa, NULL);

	while (1) {
		int full;
		client *cl, *prev_cl, *next_cl;
		fd_set read_fds, write_fds;

		do {
			sigset_t old_sigs;
			int max = -1;

			full = conn_count >= max_conns;
			FD_ZERO(&read_fds);
			fd_set_add(selfpipe[0], &read_fds, &max);
			/* If we're full, and we don't have a response to send,
			   then we don't want to accept new connections -- so
			   don't check listen_fd. */
			if (!(full && !response))
				fd_set_add(listen_fd, &read_fds, &max);

			FD_ZERO(&write_fds);
			for (cl = clients; cl; cl = cl->next)
				fd_set_add(cl->fd, &write_fds, &max);

			if (sigprocmask(SIG_UNBLOCK, &sig_chld, &old_sigs) < 0)
				die("unable to unblock SIGCHLD");

			n = select(max + 1, &read_fds, &write_fds, NULL, NULL);
			if (n < 0 && errno != EINTR)
				warn("select failed");

			if (sigprocmask(SIG_SETMASK, &old_sigs, NULL) < 0)
				die("unable to restore signal mask");
		} while (n < 0);

		if (FD_ISSET(selfpipe[0], &read_fds)) {
			char c;

			/* We don't care if this fails. */
			read(selfpipe[0], &c, 1);

			reap_children();
		}

		if (FD_ISSET(listen_fd, &read_fds)) {
			accept_connection(listen_fd, full);
		}

		prev_cl = NULL;
		for (cl = clients; cl; cl = next_cl) {
			next_cl = cl->next;

			if (FD_ISSET(cl->fd, &write_fds))
				try_to_send(prev_cl, cl);
			prev_cl = cl;
		}
	}

	return 0;
}
